File: apps/api/treinozap/entities/workout.py
from dataclasses import dataclass, field
from typing import Optional
from datetime import datetime


@dataclass
class Exercise:
    id: str
    workout_id: str
    name: str
    sets: Optional[int]
    reps: Optional[str]
    rest_seconds: Optional[int]
    notes: Optional[str]
    order_index: int
    created_at: datetime

    @classmethod
    def from_dict(cls, data: dict) -> 'Exercise':
        return cls(
            id=data['id'],
            workout_id=data['workout_id'],
            name=data['name'],
            sets=data.get('sets'),
            reps=data.get('reps'),
            rest_seconds=data.get('rest_seconds'),
            notes=data.get('notes'),
            order_index=data.get('order_index', 0),
            created_at=datetime.fromisoformat(data['created_at']),
        )


@dataclass
class Workout:
    id: str
    trainer_id: str
    student_id: Optional[str]
    title: str
    description: Optional[str]
    public_token: str
    is_active: bool
    created_at: datetime
    exercises: list[Exercise] = field(default_factory=list)

    @classmethod
    def from_dict(cls, data: dict) -> 'Workout':
        exercises = [Exercise.from_dict(e) for e in data.get('exercises', [])]
        return cls(
            id=data['id'],
            trainer_id=data['trainer_id'],
            student_id=data.get('student_id'),
            title=data['title'],
            description=data.get('description'),
            public_token=data['public_token'],
            is_active=data.get('is_active', True),
            created_at=datetime.fromisoformat(data['created_at']),
            exercises=exercises,
        )

File: packages/types/src/workout.ts
import type { UUID } from './common'

export type Exercise = {
  id: UUID
  workoutId: UUID
  name: string
  sets: number | null
  reps: string | null      // ex: "12-15" ou "falha"
  restSeconds: number | null
  notes: string | null
  orderIndex: number
  createdAt: string
}

export type ExerciseCreate = Omit<Exercise, 'id' | 'workoutId' | 'createdAt'>

export type Workout = {
  id: UUID
  trainerId: UUID
  studentId: UUID | null
  title: string
  description: string | null
  publicToken: string
  isActive: boolean
  createdAt: string
}

export type WorkoutCreate = Pick<Workout, 'trainerId' | 'studentId' | 'title' | 'description'>

export type WorkoutWithExercises = Workout & {
  exercises: Exercise[]
}

export type ActivityLog = {
  id: UUID
  studentId: UUID | null
  workoutId: UUID | null
  event: string
  metadata: Record<string, unknown> | null
  createdAt: string
}
